### PR TITLE
TRPL: associated constants

### PR DESCRIPTION
**Pull request by steveklabnik**

### CODE REVIEW

File: src/doc/trpl/associated-constants.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
+% Associated Constants
+
+With the `associated_constants` feature, you can define constants like this:
```
Comment (alexcrichton) [inline src/doc/trpl/associated-constants.md]: associated_constants => associated_consts

**Comment by alexcrichton** [comment]
I believe there's also an allowance for associated constants of the form:

```rust
pub struct Foo;
impl Foo {
    pub const FOO: u32 = 3;
}
```

e.g. an inherent constant associated to a type (maybe worth mentioning?)

**Comment by alexcrichton** [comment]
cc @quantheory

File: src/doc/trpl/associated-constants.md (hunk 1 above)
Comment (steveklabnik) [inline src/doc/trpl/associated-constants.md]: whoops!

**Comment by steveklabnik** [comment]
Yeah, worth mentioning for sure. Added, and nits fixed.

File: src/doc/trpl/associated-constants.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    assert_eq!(1, <i32 as Foo>::ID);
+    assert_eq!(5, <i64 as Foo>::ID);
+}
+```
+
+As you can see, when implementing `Foo`, you can leave it unimplemented, as
+with `i32`. It will then use the default value. But, as in `i64`, we can also
+add our own definition.
+
+Associated constants don’t have to be with a trait. An `impl` block for a `struct`
```
Comment (quantheory) [inline src/doc/trpl/associated-constants.md]: This phrasing is slightly odd. Did you mean "associated with a trait"?

**Comment by quantheory** [comment]
LGTM. There is also the ability to define a constant of type `Self`, just the bare minimum needed for cases like `Int::MAX`. But it is finicky (e.g. you can use `Self` in a trait, but have to use the exact type in an impl). I don't know if we want to describe that in the book yet, or wait for the way we deal with type parameters and associated consts to stabilize a bit.

File: src/doc/trpl/associated-constants.md (hunk 2 above)
Comment (steveklabnik) [inline src/doc/trpl/associated-constants.md]: Yes, I did.

**Comment by steveklabnik** [comment]
Yeah, let's hold off on the `Self` stuff.

File: src/doc/trpl/associated-constants.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+impl Foo for i32 {
+}
+
+impl Foo for i64 {
+    const ID: i32 = 5;
+}
+
+fn main() {
+    assert_eq!(1, <i32 as Foo>::ID);
+    assert_eq!(5, <i64 as Foo>::ID);
```
Comment (alexcrichton) [inline src/doc/trpl/associated-constants.md]: As one final nit, I think these are more idiomatically referred to without the angle-bracket syntax, e.g. `i32::ID`

File: src/doc/trpl/associated-constants.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+trait Foo {
+    const ID: i32;
+}
+
+impl Foo for i32 {
+    const ID: i32 = 1;
+}
+
+fn main() {
+    assert_eq!(1, <i32 as Foo>::ID);
```
Comment (alexcrichton) [inline src/doc/trpl/associated-constants.md]: (same angle brackets here)

**Comment by alexcrichton** [comment]
r=me with one final nit, thanks @steveklabnik!

**Pull request merged**